go/mysql/server.go

```diff
@@ -335,60 +338,39 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32, acceptTime time.Ti
 			queryStart := time.Now()
 			query := c.parseComQuery(data)
 			c.recycleReadPacket()
-			fieldSent := false
-			// sendFinished is set if the response should just be an OK packet.
-			sendFinished := false
-			err := l.handler.ComQuery(c, query, func(qr *sqltypes.Result) error {
-				if sendFinished {
-					// Failsafe: Unreachable if server is well-behaved.
-					return io.EOF
-				}
-
-				if !fieldSent {
-					fieldSent = true
-
-					if len(qr.Fields) == 0 {
-						sendFinished = true
-						// We should not send any more packets after this.
-						return c.writeOKPacket(qr.RowsAffected, qr.InsertID, c.StatusFlags, 0)
-					}
-					if err := c.writeFields(qr); err != nil {
-						return err
-					}
-				}
-
-				return c.writeRows(qr)
-			})
-
-			// If no field was sent, we expect an error.
-			if !fieldSent {
-				// This is just a failsafe. Should never happen.
-				if err == nil || err == io.EOF {
-					err = NewSQLErrorFromError(errors.New("unexpected: query ended without no results and no error"))
-				}
+			stmts, err := sqlparser.SplitStatementToSlice(query)
```

Comment:
I think this will be expensive because this work will be repeated again by vtgate. For now, it may be better to perform the split only if multi-statement is enabled. So, only those who enable it pay the price.

However, the better fix will be to change the parser to return multiple ASTs. But that will require some refactoring, which can be done later.

Reply:
I agree with sougou here. Given that you have a flag controlling support anyway, I think we should make this simple optimization to avoid splitting for the cases where multi-statement execution is disabled.